src/services/identity/UsersService.ts

```diff
@@ -230,56 +233,201 @@ class UsersService {
+              )
+
+              return new DatabaseError("Error destroying OTP entry in database")
+            }
+          ).andThen(() => errAsync(new BadRequestError("Hashed OTP not found")))
+        }
+
+        return okAsync(otpDbEntry)
+      })
+      .andThen((otpDbEntry) =>
+        // increment attempts
+        ResultAsync.fromPromise(
+          this.otpRepository.increment("attempts", {
+            where: { id: otpDbEntry.id },
+            transaction,
+          }),
+          (error) => {
+            logger.error(
+              `Error incrementing OTP attempts: ${JSON.stringify(error)}`
+            )
+
+            return new DatabaseError("Error incrementing OTP attempts")
+          }
+        ).map(() => otpDbEntry)
+      )
+      .andThen((otpDbEntry) =>
+        ResultAsync.combine([
+          okAsync(otpDbEntry),
+          ResultAsync.fromPromise(
+            this.otpService.verifyOtp(otp, otpDbEntry.hashedOtp),
+            (error) => {
+              logger.error(`Error verifying OTP: ${JSON.stringify(error)}`)
+
+              return new BadRequestError("Error verifying OTP")
+            }
+          ),
+        ])
+      )
+      .andThen(([otpDbEntry, isValidOtp]) => {
+        if (!isValidOtp) {
+          return errAsync(new BadRequestError("OTP is not valid"))
+        }
+
+        if (isValidOtp && otpDbEntry.expiresAt < new Date()) {
+          return ResultAsync.fromPromise(
+            otpDbEntry.destroy({ transaction }),
+            (error) => {
+              logger.error(
+                `Error destroying OTP entry: ${JSON.stringify(error)}`
+              )
+
+              return new DatabaseError("Error destroying OTP entry in database")
+            }
+          ).andThen(() => errAsync(new BadRequestError("OTP has expired")))
+        }
+
+        return okAsync(otpDbEntry)
+      })
+      .andThen((otpDbEntry) =>
+        // destroy otp before returning true since otp has been "used"
+        ResultAsync.fromPromise(
+          otpDbEntry.destroy({ transaction }),
+          (error) => {
+            logger.error(`Error destroying OTP entry: ${JSON.stringify(error)}`)
+
+            return new DatabaseError("Error destroying OTP entry in database")
+          }
+        )
+          .andThen(() =>
+            ResultAsync.fromPromise(transaction.commit(), (txError) => {
+              logger.error(
+                `Error committing transaction: ${JSON.stringify(txError)}`
+              )
+              return new DatabaseError("Error committing transaction")
+            })
+          )
+          .map(() => true)
+      )
+      .orElse((error) =>
+        ResultAsync.fromPromise(transaction.commit(), (txError) => {
+          logger.error(
+            `Error committing transaction: ${JSON.stringify(txError)}`
+          )
+          return new DatabaseError("Error committing transaction")
+        }).andThen(() => errAsync(error))
+      )
```

Comment:
there's honestly alot of repeated `fromPromise` -> error handling code here. 

this is fine, but i suspect the error handling code is too verbose and takes away from the actual business logic of what we're trying to do here. 

maybe a better way to do it would be to write a wrapper for `fromPromise` that is localised to this module and does a simple log of the error + returns a `DatabaseError` (cos everything here seems to return that) 

don't have to do it now, just food for thought